src/invidious/videos/storyboard.cr

```diff
@@ -0,0 +1,119 @@
+require "uri"
+require "http/params"
+
+module Invidious::Videos
+  struct Storyboard
+    # Template URL
+    getter url : URI
+    getter proxied_url : URI
+
+    # Thumbnail parameters
+    getter width : Int32
+    getter height : Int32
+    getter count : Int32
+    getter interval : Int32
+
+    # Image (storyboard) parameters
+    getter rows : Int32
+    getter columns : Int32
+    getter images_count : Int32
+
+    def initialize(
+      *, @url, @width, @height, @count, @interval,
+      @rows, @columns, @images_count
+    )
+      authority = /(i\d?).ytimg.com/.match(@url.host.not_nil!).not_nil![1]?
+
+      @proxied_url = URI.parse(HOST_URL)
+      @proxied_url.path = "/sb/#{authority}#{@url.path}"
+      @proxied_url.query = @url.query
+    end
+
+    # Parse the JSON structure from Youtube
+    def self.from_yt_json(container : JSON::Any, length_seconds : Int32) : Array(Storyboard)
+      # Livestream storyboards are a bit different
+      # TODO: document exactly how
```

Comment:
Tried messing around with them in shaka for FreeTube (as we can ise the live dash manifest with shaka and get 4 hours of seeking) but the code I came up with based on reversing the player.js file and stepping through it with a debugger, ended up requesting loads of future not yet existing thumbnails (something in my assumptions or calculations must have been off), so I gave up and continued doing the rest of the shaka migration.
I'll let you know if I ever get it working in FreeTube.

Reply:
Ok, thanks for your research ^^

Reply:
Implementing live storyboards is probably something better done on the client side anyway, as it will require information from the streaming URL response headers or the live DASH manifest (at least that's the only way I can think of figuring out the current live segment) and you'll have to update it every 5 seconds, so generating a VTT file on the server side doesn't seem like a viable option for that.

My previous attempt revolved around using shaka's manifest preprocessing function to inject the live storyboards into the DASH manifest based on information in the manifest. Although as shaka already struggles enough with YouTube's 4 hour DVR window (4 hours worth of 5 second segments = 2880, which it has to recalculate every 5 seconds x) ) in the live manifests, so I'm not sure if adding that extra delay is a good idea, even if my solution had worked, which it didn't.